### PR TITLE
Support for org.opentest4j.FileInfo

### PR DESCRIPTION
**Pull request by donat**
Fixes https://github.com/gradle/gradle/issues/24801

### Todo
- [x] JUnit 4 support
- [x] JUnit 5 support
- [x] Expose file content
- [x] Expose file paths

### CODE REVIEW

File: subprojects/build-events/src/main/java/org/gradle/internal/build/event/types/DefaultFileComparisonTestAssertionFailure.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import org.gradle.tooling.internal.protocol.InternalFailure;
+import org.gradle.tooling.internal.protocol.InternalFileComparisonTestAssertionFailure;
+
+import java.util.Collections;
+import java.util.List;
+
+public class DefaultFileComparisonTestAssertionFailure extends DefaultTestAssertionFailure implements InternalFileComparisonTestAssertionFailure {
+
+    private final byte[] expectedContent;
+    private final byte[] actualContent;
```
Comment (hegyibalint) [inline subprojects/build-events/src/main/java/org/gradle/internal/build/event/types/DefaultFileComparisonTestAssertionFailure.java:27]: 🤔 Is storing the entire file in memory a good choice?
I'm sure somebody in the universe will have a GB/TB-sized file for some reason, and you will need to keep it in the memory.

File: subprojects/architecture-test/src/changes/archunit_store/internal-api-nullability.txt [hunk 2]
```diff
@@ -882,7 +882,6 @@ Class <org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor$Method
 Class <org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor> is not annotated (directly or via its package) with @org.gradle.api.NonNullApi in (JUnitTestClassExecutor.java:0)
 Class <org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor> is not annotated (directly or via its package) with @org.gradle.api.NonNullApi in (JUnitTestClassProcessor.java:0)
 Class <org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessorFactory> is not annotated (directly or via its package) with @org.gradle.api.NonNullApi in (JUnitTestClassProcessorFactory.java:0)
-Class <org.gradle.api.internal.tasks.testing.junit.JUnitTestEventAdapter> is not annotated (directly or via its package) with @org.gradle.api.NonNullApi in (JUnitTestEventAdapter.java:0)
```
Comment (hegyibalint) [inline subprojects/architecture-test/src/changes/archunit_store/internal-api-nullability.txt:885]: Made the work to use `NonNullApi` on the class.

File: subprojects/architecture-test/src/changes/archunit_store/internal-api-nullability.txt [hunk 3]
```diff
@@ -945,7 +944,6 @@ Class <org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClas
 Class <org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor> is not annotated (directly or via its package) with @org.gradle.api.NonNullApi in (JUnitPlatformTestClassProcessor.java:0)
 Class <org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor> is not annotated (directly or via its package) with @org.gradle.api.NonNullApi in (JUnitPlatformTestClassProcessor.java:0)
 Class <org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessorFactory> is not annotated (directly or via its package) with @org.gradle.api.NonNullApi in (JUnitPlatformTestClassProcessorFactory.java:0)
-Class <org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestExecutionListener> is not annotated (directly or via its package) with @org.gradle.api.NonNullApi in (JUnitPlatformTestExecutionListener.java:0)
```
Comment (hegyibalint) [inline subprojects/architecture-test/src/changes/archunit_store/internal-api-nullability.txt:948]: Made the work to use `NonNullApi` on the class.

File: subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java [hunk 4]
```diff
@@ -124,70 +144,20 @@ private void reportTestFailure(TestIdentifier testIdentifier, Throwable failure)
...
-                return null;
+    public TestFailure createFailure(Throwable failure) {
+        for (FailureMapper mapper : MAPPERS) {
+            if (mapper.supports(failure.getClass())) {
+                try {
+                    return mapper.map(failure, this);
+                } catch (Exception ignored) {
+                    // TODO before merge: Issue to do debug logging?
+                    // ignore
+                    System.out.println(ignored.getMessage());
```
Comment (donat) [inline subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java]: ❌ I'm fine with simply ignoring the exception here if it was the case previously.

File: subprojects/testing-jvm-infrastructure/build.gradle.kts [hunk 5]
```diff
@@ -41,6 +41,16 @@ dependencies {
...
+    testImplementation(libs.assertj) {
+        because("We test assertion errors coming from AssertJ")
+    }
+    testImplementation("org.opentest4j:opentest4j") {
+        version {
+            // MultipleFailuresError appears only since 1.3.0-RC2
+            require("1.3.0-RC2")
+        }
+        because("We test assertion errors coming from OpenTest4J")
+    }
```
Comment (donat) [inline subprojects/testing-jvm-infrastructure/build.gradle.kts:53]: ❓ My understanding is that the Gradle code only reflectively interacts with opentest4j classes. Why do we need the dependency here?

File: subprojects/testing-jvm-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/failure/FailureMapper.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+import org.gradle.api.NonNullApi;
+import org.gradle.api.tasks.testing.TestFailure;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.List;
+
+/**
+ * A mapper that maps a {@link Throwable} thrown from a test engine into a {@link TestFailure}.
```
Comment (donat) [inline subprojects/testing-jvm-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/failure/FailureMapper.java:27]:
```suggestion
 * A mapper that maps a {@link Throwable} thrown during test execution into a {@link TestFailure}.
```

File: subprojects/testing-jvm-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/failure/FailureMapper.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+     *
+     * @param throwable the {@link Throwable} to be mapped
+     * @param rootMapper the {@link RootAssertionToFailureMapper} to be used to recursively map inner failures.
+     * @return a {@link TestFailure} representing the given {@link Throwable}
+     * @throws Exception if an error occurs while mapping the {@link Throwable}
+     */
+    public abstract TestFailure map(Throwable throwable, RootAssertionToFailureMapper rootMapper) throws Exception;
+
+    // Utility methods ------------------------------------
+
```
Comment (donat) [inline subprojects/testing-jvm-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/failure/FailureMapper.java:81]:
```suggestion
```

File: subprojects/testing-jvm-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/failure/RootAssertionToFailureMapper.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+ * As sometimes failure processing needs to be recursive (e.g. {@link org.gradle.api.internal.tasks.testing.failure.mappers.OpentestMultipleFailuresMapper}),
+ * they receive an instance implementing this interface to transform inner failures.
+ *
+ * @see org.gradle.api.internal.tasks.testing.junit.JUnitTestEventAdapter
+ * @see org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestExecutionListener
+ */
+@NonNullApi
+public interface RootAssertionToFailureMapper {
+
+    TestFailure createFailure(Throwable throwable);
```
Comment (donat) [inline subprojects/testing-jvm-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/failure/RootAssertionToFailureMapper.java:34]: 🤔 Would it make sense to return the list of mapper instances here?

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalFileComparisonTestAssertionFailure.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+     * Todo.
+     */
+    @Nullable
+    byte[] getExpectedContent();
+
+    /**
+     * Todo.
+     */
+    @Nullable
+    byte[] getActualContent();
```
Comment (donat) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalFileComparisonTestAssertionFailure.java:44]: ❌  three missing TODOs

**Review by donat — COMMENTED** [review]
image

File: subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java (hunk 4 above)
Comment (hegyibalint) [reply subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java]: PTAL again, now this uses a logger.
I'm on a fence on even ignoring this exception, as with the new mappers, you should not have a case when a mapper fail.

File: subprojects/testing-jvm-infrastructure/build.gradle.kts (hunk 5 above)
Comment (hegyibalint) [reply subprojects/testing-jvm-infrastructure/build.gradle.kts:53]: This is `testImplementation`; we instantiate a failure in the tests.
https://github.com/gradle/gradle/blob/1fdeaddce73e2884d783a2e138e1b7897ad75713/subprojects/testing-jvm-infrastructure/src/test/groovy/org/gradle/api/internal/tasks/testing/failure/mappers/OpentestMultipleFailuresMapperTest.groovy#L30-L37

File: subprojects/testing-jvm-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/failure/RootAssertionToFailureMapper.java (hunk 8 above)
Comment (hegyibalint) [reply subprojects/testing-jvm-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/failure/RootAssertionToFailureMapper.java:34]: If your idea is to only supply the list of `FailureMapper` classes here: I think that exactly against the flexibility we talked about some time ago.

That would mean that you wouldn't have a chance to do non-mapper-based logic e.g.:
https://github.com/gradle/gradle/blob/2a32ceedf4406e782b697af9b396f2fcea7e7f9d/subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java#L151-L163
Currently, if you need some custom logic (something, which doesn't fit the mappers' concepts), you could just write it in this method; as it's used recursively, you wouldn't need to crunch those custom parts into mappers.

Comment (donat) [reply subprojects/testing-jvm-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/failure/RootAssertionToFailureMapper.java:34]: That makes sense, thanks for the explanation.

File: subprojects/testing-jvm-infrastructure/build.gradle.kts (hunk 5 above)
Comment (donat) [reply subprojects/testing-jvm-infrastructure/build.gradle.kts:53]: Ah, my bad. All clear now.

File: subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java (hunk 4 above)
Comment (donat) [reply subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java]: This should never happen, so reporting an error here makes sense to me.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalFileComparisonTestAssertionFailure.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+import org.gradle.api.Incubating;
+import org.gradle.api.NonNullApi;
+
+import javax.annotation.Nullable;
+
+/**
+ * DO NOT CHANGE THIS INTERFACE. It is part of the cross-version protocol.
+ *
+ * Todo.
```
Comment (donat) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalFileComparisonTestAssertionFailure.java:26]: ❌

**Comment by hegyibalint** [comment]
@bot-gradle test and merge please

File: subprojects/tooling-api/src/testFixtures/groovy/org/gradle/integtests/tooling/fixture/TapiBuildScript.groovy [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.tooling.fixture
+
+class TapiBuildScript {
```
Comment (big-guy) [inline subprojects/tooling-api/src/testFixtures/groovy/org/gradle/integtests/tooling/fixture/TapiBuildScript.groovy:19]: Is this used for anything?

**Pull request merged**